[ty] synthesize __replace__ for >=3.13

thejchap opened the pull request:
## Summary
https://github.com/astral-sh/ty/issues/111

adds support for the new `copy.replace` and `__replace__` protocol [added in 3.13](https://docs.python.org/3/whatsnew/3.13.html#copy)

## Test Plan
TODO

thejchap commented on crates/ty_python_semantic/src/types/class.rs:
```diff
@@ -1659,21 +1661,27 @@ impl<'db> ClassLiteral<'db> {
+                // When replacing, we know there is a default value for the field
+                // (the value that is currently assigned to the field)
+                // assume this to be the declared type of the field
+                if for_replace {
```
not sure this is the right way to model this...

```py
reveal_type(a.__replace__)

# mypy: def (*, a: builtins.int =) -> test.Test
# ty: (*, a: int = int) -> Test
```

sharkdp replied:
The type of the default value is not that important, and we're even thinking about removing this from the displayed signature (showing an actual value instead — if available. If none is available, like here, we could still display `= ...`). So what you have here seems fine. I would move the comment inside the `if` branch, though.

thejchap commented on crates/ty_python_semantic/src/types/class.rs:
```diff
@@ -1592,7 +1592,9 @@ impl<'db> ClassLiteral<'db> {
-        let signature_from_fields = |mut parameters: Vec<_>| {
+        let signature_from_fields = |mut parameters: Vec<_>,
+                                     return_ty: Option<Type<'db>>,
+                                     for_replace: bool| {
```
didn't feel like the cleanest way to toggle some of this behavior, open to suggestions...

thejchap commented:
@AlexWaygood ready for some feedback on this one!

sharkdp commented on crates/ty_python_semantic/resources/mdtest/replace.md:
````diff
@@ -0,0 +1,64 @@
+
+It is present in the `copy` module.
+
+```toml
+[environment]
+python-version = "3.13"
+```
````
If you move this block up to the top level `# Replace` section, it will be active for all sub-sections. I think it makes sense to do this here, as every test will otherwise need to have one of these.

sharkdp commented on crates/ty_python_semantic/src/types/class.rs:
```diff
@@ -1735,6 +1743,20 @@ impl<'db> ClassLiteral<'db> {
                     .place
                     .ignore_possibly_unbound()
             }
+            (CodeGeneratorKind::DataclassLike, "__replace__") => {
+                if Program::get(db).python_version(db) < PythonVersion::PY313 {
+                    return None;
+                }
```
```suggestion
            (CodeGeneratorKind::DataclassLike, "__replace__") if Program::get(db).python_version(db) < PythonVersion::PY313 => {
```

sharkdp reviewed (approved):
Very nice — thank you. I added a couple of minor inline comments.

thejchap commented:
@sharkdp thanks! addressed most of your comments, and added a wip implementation of special handling of `copy.replace` [here](https://github.com/astral-sh/ruff/pull/19545/files#diff-89ab607f2738327350b89c4bf98de6e4af15eb4cfdf55d3afc1cb71499928f88R952). please let me know if that's on the right track

if so, i'm fine to either merge as-is and complete that in a follow-on pr, or finish as part of this one

sharkdp reviewed (approved):
Thank you very much. I made a few minor adjustments.
